core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -516,6 +516,7 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     _applicationId = _taskScheduler.applicationId()
     _applicationAttemptId = taskScheduler.applicationAttemptId()
     _conf.set("spark.app.id", _applicationId)
+    _ui.foreach(_.setAppId(_applicationId))
```

Comment:
This is actually the only change that seems necessary to fix the problem. All the other changes could be reverted, I think.